crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,495 @@
+    type Amount = U256;
+    type Commitment = (F, F, F);
+    type LookupProof = ();
+    // TODO(Chengyu): Can we make it references?
+    type IntoIter = <Vec<(K1, U256, K2)> as ark_std::iter::IntoIterator>::IntoIter;
+
+    fn register(
+        &mut self,
+        new_key: Self::Key,
+        amount: Self::Amount,
+        aux: Self::Aux,
+    ) -> Result<(), StakeTableError> {
+        match self.bls_mapping.get(&new_key) {
+            Some(_) => Err(StakeTableError::ExistingKey),
+            None => {
+                let pos = self.bls_mapping.len();
+                self.head.bls_keys.push(new_key.clone());
+                self.head.schnorr_keys.push(aux);
+                self.head.stake_amount.push(amount);
+                self.head_total_stake += amount;
+                self.bls_mapping.insert(new_key, pos);
+                Ok(())
+            }
+        }
+    }
+
+    fn deregister(&mut self, existing_key: &Self::Key) -> Result<(), StakeTableError> {
+        match self.bls_mapping.get(existing_key) {
+            Some(pos) => {
+                self.head_total_stake -= self.head.stake_amount[*pos];
+                self.head.stake_amount[*pos] = U256::zero();
+                Ok(())
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    fn commitment(&self, version: SnapshotVersion) -> Result<Self::Commitment, StakeTableError> {
+        match version {
+            // IMPORTANT: we don't support committing the head version b/c it's not finalized.
+            SnapshotVersion::Head => Err(StakeTableError::SnapshotUnsupported),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_comm),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_comm),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn total_stake(&self, version: SnapshotVersion) -> Result<Self::Amount, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head_total_stake),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start_total_stake),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start_total_stake),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+
+    fn len(&self, version: SnapshotVersion) -> Result<usize, StakeTableError> {
+        Ok(self.get_version(version)?.bls_keys.len())
+    }
+
+    fn contains_key(&self, key: &Self::Key) -> bool {
+        self.bls_mapping.contains_key(key)
+    }
+
+    fn lookup(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<Self::Amount, StakeTableError> {
+        let table = self.get_version(version)?;
+        let pos = self.lookup_pos(key)?;
+        if pos >= table.bls_keys.len() {
+            Err(StakeTableError::KeyNotFound)
+        } else {
+            Ok(table.stake_amount[pos])
+        }
+    }
```

Comment:
I have a slight concern w/ this.
the `self.bls_mapping` is not per-snapshot, which means if our deregistering will prune entries with zero amount, then the `pos` of a certain bls key might be shifted. 

This logic is only correct if we assume ever-growing vectors for `Snapshot::bls_keys/schnorr_keys/amounts` (even the non-zero staker is always bounded at 1000, the length of the vector will be growing)

this will lead to some awkward memory management issue imo.

Reply:
I assume this problem is in couple of whether we remove the entry when deregistering the key. 
I would suggest we keep everything as is for now (we simply set the stake amount to zero without actual deregistration). And figure out a better way in the future.

Reply:
Let me paste another related thread here, for easier future back reference:

> [alxiong](https://github.com/alxiong)(https://github.com/EspressoSystems/HotShot/pull/1932#discussion_r1385291474)
> Some concerns:
>
> since during deregister, we never remove that entry, that key will be included in this commitment computation
similarly, the stake table will never be truly "full" (at 1000 stakers) due to withdrawn stakers taking up the space/indexes.
> 
> @[mrain](https://github.com/mrain)(https://github.com/EspressoSystems/HotShot/pull/1932#discussion_r1385369341)
>Even if you remove the entry, in the commitment computation some default value will take that place.
For commitment computation, I'm padding the stake table to its capacity considering we need to do it in a circuit

Reply:
I understand the "padding for circuit purpose" part, Also, I don't mind the "zero-balance pk included in the stake table commitment computation" aspect that much for now (altho it incur some useless hashing)

but I'm pointing out the fact that imagine we cap the table size to be 1000, there are 900 registered then deregistered, now, we only have 100 slots left for actual active stakers.

This is a problem.

(I don't mind merging this PR now and dealing with this in a separate issue, but need acknowledgement from @jbearer @chancharles92 and @philippecamacho)

Reply:
a naive approach is: 
- when deregister, remove the entry from the stake table vector (not just mark balance to 0)
- keep an extra `pruned_indices: Vec<usize>` list of pruned positions, so the next register will fill up these first
- to enable `lookup` on `EpochStart` and `LastEpochStart` snapshots of the stake table, keep another 2 super small HashMap, of the pruned bls_key and their original pos. 


another approach with slightly lower overhead on state management is:
- keep 3 (small) lists: `pruned_indices_cur_epoch`,  `pruned_indices_last_epoch`, and `pruned_indices_sec_last_epoch` (maybe think of better names)
- during register, instead of immediately filling up all available pruned positions, we only fill up those in second last epoch (i.e. 2 epochs ago), if there are more registration, we simply append to the current head.
  - this obviate the need for the 2 small hashmaps to enable efficient lookup on the 3 snapshots we want to support.

Reply:
I had this garbage collection idea before, but never filled in the details. Let's do it in another PR (in the future)

Reply:
> I understand the "padding for circuit purpose" part, Also, I don't mind the "zero-balance pk included in the stake table commitment computation" aspect that much for now (altho it incur some useless hashing)
> 
> but I'm pointing out the fact that imagine we cap the table size to be 1000, there are 900 registered then deregistered, now, we only have 100 slots left for actual active stakers.
> 
> This is a problem.
> 
> (I don't mind merging this PR now and dealing with this in a separate issue, but need acknowledgement from @jbearer @chancharles92 and @philippecamacho)

I am ok as well to address this issue later.